lib/jxl/enc_frame.cc

```diff
@@ -459,6 +460,199 @@ void SimplifyInvisible(Image3F* image, const ImageF& alpha, bool lossless) {
+  FrameInfo patch_frame_info;
+  patch_frame_info.save_as_reference = 2;
+  patch_frame_info.frame_type = FrameType::kReferenceOnly;
+  patch_frame_info.save_before_color_transform = true;
+
+  ImageBundle ib(&state->shared.metadata->m);
+  patch_frame_info.ib_needs_color_transform = false;
+
+  ib.SetFromImage(std::move(nonphoto),
+                  state->shared.metadata->m.color_encoding);
+  if (!ib.metadata()->extra_channel_info.empty()) {
+    // Add dummy extra channels to the patch image: patches do not yet support
+    // extra channels, but the codec expects that the amount of extra channels
+    // in frames matches that in the metadata of the codestream.
+    std::vector<ImageF> extra_channels;
+    extra_channels.reserve(ib.metadata()->extra_channel_info.size());
+    for (size_t i = 0; i < ib.metadata()->extra_channel_info.size(); i++) {
+      extra_channels.emplace_back(ib.xsize(), ib.ysize());
+      FillImage(0.0f, &extra_channels.back());
+    }
+    ib.SetExtraChannels(std::move(extra_channels));
+  }
+
+  PassesEncoderState roundtrip_state;
+  auto special_frame = std::unique_ptr<BitWriter>(new BitWriter());
+  JXL_CHECK(EncodeFrame(cparams, patch_frame_info, state->shared.metadata, ib,
+                        &roundtrip_state, pool, special_frame.get(), nullptr));
+  const Span<const uint8_t> encoded = special_frame->GetSpan();
+  state->special_frames.emplace_back(std::move(special_frame));
+
+  if (true ||
+      cparams.butteraugli_distance < kMinButteraugliToSubtractOriginalPatches) {
+    BitReader br(encoded);
+    ImageBundle decoded(&state->shared.metadata->m);
+    PassesDecoderState dec_state;
+    JXL_CHECK(dec_state.output_encoding_info.Set(
+        *state->shared.metadata,
+        ColorEncoding::LinearSRGB(
+            state->shared.metadata->m.color_encoding.IsGray())));
+    JXL_CHECK(DecodeFrame({}, &dec_state, pool, &br, &decoded,
+                          *state->shared.metadata, /*constraints=*/nullptr));
+    // if the modular nonphoto frame uses patches, we need to decode another
+    // frame
+    if (!dec_state.shared_storage.reference_frames[2].storage.color()->xsize())
+      JXL_CHECK(DecodeFrame({}, &dec_state, pool, &br, &decoded,
+                            *state->shared.metadata, /*constraints=*/nullptr));
+
+    JXL_CHECK(br.Close());
+    state->shared.reference_frames[2] =
+        std::move(dec_state.shared_storage.reference_frames[2]);
+  } else {
+    state->shared.reference_frames[2].storage = std::move(ib);
+  }
+  state->shared.reference_frames[2].frame =
+      &state->shared.reference_frames[2].storage;
+
+  for (size_t c = 0; c < 3; ++c) {
+    for (size_t y = 0; y < ysize; y++) {
+      float* JXL_RESTRICT p = opsin->PlaneRow(c, y);
+      const float* JXL_RESTRICT np =
+          state->shared.reference_frames[2].frame->color()->ConstPlaneRow(c, y);
+      for (size_t x = 0; x < xsize; x++) {
+        p[x] -= np[x];
+      }
+    }
+  }
```

Comment:
Most of the code above seems duplicated from the traditional patch detection, can we extract that as a function?